### DISCUSSION TITLE
Cache interface options

### DISCUSSION THREAD

**Discussion by andrewhill157** [marimo-team/marimo#2653] [Q&A]
I was looking more closely into the cache options added recently and saw that you'd added `mo.cache` and `mo.persistent_cache`! Sorry if this overlaps with past discussion that I've missed somewhere!

My understanding is that the former is used like
```
@mo.cache
def my_function():
    sleep(5)
    return 0
```

whereas the latter can be used like:
```
with mo.persistent_cache("my_cache"):
    x = expensive_thing()
```

but can't be used to decorate functions and the docs point you to use of `@mo.cache` for that.

In looking around at other options I came across [joblib.Memory](https://joblib.readthedocs.io/en/latest/memory.html) which can be used to decorate functions like `@mo.cache` (or the various functools cache options) like:
```
from joblib import Memory
mem = Memory("my_cache")

@mem.cache
def my_function():
    sleep(5)
    return 0
```

but with a persistent cache.

I don't think you can directly use it as a context manager but you could obvs make something to do that pretty easily.

One thing I liked about this was that it keeps the interface when using a cache vs. a persistent cache very similar. Not sure if you would (or already have) consider anything like this as an alternative or extension of the ways marimo currently supports caching.

`joblib.Memory` works fine for my purposes if decorating a function defined in a module outside of a notebook, but it doesn't seem to persistently cache if the function is defined inside the notebook. Seems like that is probably a known issue:
https://github.com/joblib/joblib/issues/1498

**Comment by akshayka** [comment]
@andrewhill157 , thanks for opening this discussion and so sorry for the late response. 

You're correct that we have an API asymmetry between `mo.cache` and `mo.persistent_cache`, one is a decorator and the other a context manager.

> One thing I liked about this was that it keeps the interface when using a cache vs. a persistent cache very similar. Not sure if you would (or already have) consider anything like this as an alternative or extension of the ways marimo currently supports caching.

I see the appeal of having the same interface for both. cc @dmadisetti — is that something that could be possible in the future?

Separately — Andrew, I'm curious if you've gotten a chance to use our caching features, and if you had any feedback

**Reply by dmadisetti** [reply]
This would be very easy to implement, they share the same underlying API (we'd just swap out `MemoryLoader` for `PickleLoader`); the hardest part is just thinking how to expose the functionality.

I think mixing the namespace to be a decorator and a context manager will make this a little difficult from a maintenance perspective and might be a little bit confusing in documentation. But maybe there's a certain elegance to it? Others have talked about wanting other save formats than pickle as well, so I think an API extension is due. With a shared namespace we might have:

```python
from marimo import memory_cache, pickle_cache, numpy_cache

@memory_cache
def fn():
    ...

@remote_cache(endpoint="example.api")
def cached():
    ...

with numpy_cache("expensive_block", compression=True):
    ...

```

For implementation, I see this working by switching on the constructor arguments types- which is messy but maybe fine.
Conversely, 

```python
from marimo.save import MemoryLoader, RemoteLoader, NumpyLoader

# `persistent_cache` nomenclature may be misleading if the Loader doesn't actually write to disk
with MemoryLoader.context("my_cache") as cache:
    x = expensive_thing()


@RemoteLoader.cache(endpoint="example.api")
def fn():
    ...

with NumpyLoader.context("expensive_block", compression=True):
    ...
```

open to more thoughts too

**Reply by akshayka** [reply]
> Others have talked about wanting other save formats than pickle as well, so I think an API extension is due.

Do we need to expose other save formats, or can we just handle this for users behind-the-scenes? The latter would be my preference. From the code sample it's not clear to me what `numpy_cache` would do (only cache NumPy arrays, and nothing else? Not sure we need that).

**Reply by dmadisetti** [reply]
`npy` and `npz` can save basic structures, just given as an example here, because npz has better compression than pickle. But there are other formats like dill, which are more expansive than pickle.

Choosing a file format behinds the scenes could lead to unexpected cache misses on dependency additions or "behind the scenes changes". I think it's OK to be explicit and let the user choose the save format, knowing the limitations of the chosen format.

For instance, a parquet saver would be great for saving dataframes, but won't work for general python objects

marimo could try to Frankenstein multiple save types together, but I think this makes the system more opaque and brittle. Allowing the user to choose the format type based on its strength (or weakness), and optimize their code for the given format, seems like best option

But we should still keep the general `cache` or `persistent_cache` for general cases

But open for discussion- how would a behind-the-scenes format best work?

**Reply by akshayka** [reply]
> But we should still keep the general cache or persistent_cache for general cases

Yea that makes sense.

I think I'm just concerned about API scope creep. But I do see of course why more specific serialization types would be helpful.

I'm also a bit confused by the semantics. For

```
with NumpyLoader.context("expensive_block", compression=True):
    ...
 ```
 
 if the code in `...` created variables of several types, some of which were not serializable as `.npy`, would a `RuntimeError` be raised?

**Reply by dmadisetti** [reply]
Yes, which is the same as pickle right now

**Comment by andrewhill157** [comment]
@akshayka sorry for the late reply -- I've only used them in some simple test examples so not sure I have much feedback at this point other than the interface detail and some downstream limitations (or user error).

For example, in my case the function I want to cache is called within another function, which doesn't seem to play nice with marimo's implementation currently -- for example:

```
import marimo as mo
import time

def test(i):
    with mo.persistent_cache("my_cache"):
        time.sleep(5)
        result = 5
    return result * i

test(1)
```

yields a BlockException `marimo._save.ast.BlockException: persistent_cache cannot be invoked within a block (try moving the block within the persistent_cache scope).`

I also tried to make my own decorator but run into the same error when doing so.

I don't currently have particular requirements re: save formats (pickle is fine for current use case), but like the idea of maintaining a generic (or auto) default personally.

**Reply by dmadisetti** [reply]
So this is an intentional check, since we can't get the same guarantees of the execution path if the block is called from somewhere like in a loop.

My suggestion for something like this would be:

```python
with persistent_cache("precomputed"):
    data = [expensive_call(i) for _ in range(i)]
```

Or just 

```python
with persistent_cache(f"compute_{i}"):
    data = expensive_call(i)
```

which will call on a change to `i`.

I get the use case, that you are calling `expensive_call` many times, in a variety of situations, so it doesn't fit into this format.
I'm still open to mixing the namespace, but I do want to leave room for other caching types

`persistent_cache` will likely not work with any user level decorator by design- if this is really pressing, you can swap out `MemoryLoader` for `PickleLoader` in `marimo._save._cache_base`- but I think it will be an easy feature to add once a api is agreed.

---

There is also this spec:

```python
with mo.persistent_cache("block", loader=MemoryLoader) as cache:
    ...

@mo.cache(loader=PickleLoader):
def fn():
    ...
```

```python
with mo.persistent_cache("block", loader="memory") as cache:
    ...

@mo.cache(loader="pickle") # This is what you would want
def fn():
    ...
```

Which I'm fine with, except that `persistent_cache` is potentially a misnomer (with particular loaders).
So maybe, keep persistent_cache behavior the same, and have a different function for explicit loaders:

```python
with context_cache("block", loader="parquet"):
    ...
```

**Reply by akshayka** [reply]
> @mo.cache(loader="pickle") # This is what you would want

Would this have the same semantics as `persistent_cache`?

> So maybe, keep persistent_cache behavior the same, and have a different function for explicit loaders:

That makes sense, though what happens if some of the objects saved/loaded are not compatible with the selected loader? That's on the user I guess?

**Reply by dmadisetti** [reply]
Hash names should be the same, but I think this should be on the user

**Reply by akshayka** [reply]
Makes sense

**Reply by gabrielgrant** [reply]
Just flagging that @leventov's [two proposals in the comment below](https://github.com/marimo-team/marimo/discussions/2653#discussioncomment-11464871) are relevant to this thread (I'd missed them when reading this and had started writing basically the same things)

**Comment by leventov** [comment]
As a general meta-note, I think would be nice to have a clearer conceptual picture about what we want from the notebook source files. Not just for developers, but more generally for users of marimo, and documented.

One simple distinction from Jupyter is already obvious from the system premise, that is, in virtue of being `.py` files, marimo notebooks don't store cell results. But there is still a fork on the road after that: do we want it to mostly "specification of behavior" and deployment concerns shoveled elsewhere, or we want it to bundle information about its own deployment (and caching, especially remote, is kinda a deployment concern), too. And a third option is perhaps to make it work both ways.

I don't think this is an easy question. There are good arguments that could be made for each option.

This question also interplays with another important API design dimension, namely the question of timeliness of cells (what does it mean for cells to have a time parameter, e.g. defaulting to now() but could be pegged for illustration?), and which is also an important deployment concern, if cells to be deployed in a kind of data processing DAG or executed at certain intervals.

**Comment by leventov** [comment]
Few thoughts:

1) I think there is probably little reason to configure storage formats per cell within a notebook. That is, if I want to use npy or Arrow or whatever compressed format for columns for storage, this expands to all cells within a notebook outputting columnar data. It's because the choice is motivated by space/network bandwidth usage -- load/store speed tradeoff, and this extends to the entire notebook as it's the unit of deployment.

So, maybe create App can be configured with Cacher (which in turn can be configured with different formats), and then there is `@app.persist_results` decorator. Indirection through App to save a cacher = Cacher() line if the user doesn't change the defaults.

2) Regarding specific formats, maybe use the approach of ORMs and other similar serialisers that permit configuring serialiser per type. That is, "numpy" as a kind of top level Cache doesn't make sense - it should be configured just for specific object types, whenever they appear. Others may default to pickle or whatever. So, it's more like `Cacher((np.array, "npy"), (pd.dataframe, "arrow"))`, than `NumpyCacher()`.

**Reply by leventov** [reply]
Actually, if `Cacher` is an App's field, `@app.cell(persist_results=true)` may be simpler, no need for a separate decorator.

**Reply by gabrielgrant** [reply]
Re 2. -- to configure a set of serialization methods to be used together (eg selecting different ones for different types of objects), an alternative API would be something more like [the `json` module's approach of a user-provided  "dispatcher" function that takes the object to be serialized and can choose to do something custom with it, or just fall back to the default implementation](https://docs.python.org/3/library/json.html#json.JSONEncoder.default). While the `json` dispatch function converts the object to a JSON-serializable value, in the marimo caching case, the return side of the API is a little less clear. Return a serializer function? Do the serialization and return a buffer or stream?


re: 1. I tend to agree that configuring a set of serializers for a whole notebook seems like the default use-case -- not sure if there's a realistic use-case for wanting to change this on a per-cell basis? perhaps to output some specific results in a particular format to be used in some other system? but just dumping explicitly on re-calc seems like a better approach for that. in any case, per-cell/`@cache` call certainly seems less common than just wanting it configured notebook-wide.

**Reply by dmadisetti** [reply]
I do think a default loader on a notebook level makes sense in various contexts, leveraging a general `persistent_cache` seems the best way to make it override-able at the notebook or user setting.

On a per object basis, this definitely increases the complexity for marimo- I think it's OK to rely on out of the box functionality, and let the user handle what should be explicitly saved in what format, e.g.

```python
with persistent_cache("big df", method="arrow"):
     my_df = compute_df()

with persistent_cache("np", method="numpy"):
     my_arr = compute_np()
```

Opposed to something like:

```
with Loader("cache", (np.array, "npy"), (pd.dataframe, "arrow"))
     my_df = compute_df()
     my_arr = compute_np()
```
There are tons of great serializations libraries out there, and being careful and intentional around the blocks you choose to cache and the correct method, will always out perform a general approach. I think that doing this under the hood may make sense, but it should be still over-ridable if desired (e.g. pickle is surprisingly fast, but does use more memory than other serialization methods)

**Reply by akshayka** [reply]
I actually like your second example more. I imagine for unspecified types it would use a default serialization method.

```python
with Loader("cache", (np.array, "npy"), (pd.dataframe, "arrow"))
     my_df = compute_df()
     my_arr = compute_np()
     my_dict = ... # uses pickle by default
```

> let the user handle what should be explicitly saved in what format, e.g.

Specifying multiple types is still explicit, no?

> On a per object basis, this definitely increases the complexity for marimo

API complexity or implementation? Saving would be straightforward, but I suppose loading would be less so.

**Reply by dmadisetti** [reply]
Actually wouldn't be that bad as special loader

```python
class MultiLoader(BasePersistenceLoader):
    """Loader for different types of serializable objects."""

    def __init__(self, name: str, save_path: str, lookup: dict[Type: str]) -> None:
        super().__init__(name, "multi.json", save_path)
        # Base loader to hold metadata
        self._loader = JsonLoader(f"{name}.multi", save_path=save_path)
        self.lookup = lookup

    def load_persistent_cache(
        self, hashed_context: str, cache_type: CacheType
    ) -> Cache:
        self._loader._save_path = self._save_path
        cache = self._loader.load_persistent_cache(hashed_context, cache_type)
        for loader, defs in cache.defs.items():
              obj_cache = PERSISTENT_LOADERS[loader](loader, self.save_path).load(hashed_context, cache_type)
              cache.update(obj_cache)
              # Check defs actually loaded
              assert defs == cache.def.keys()
        return cache

    def save_cache(self, cache: Cache) -> None:
        collections = {}
        for var, obj in cache.defs.items():
            loader = self.lookup.get(type(obj), "pickle")
            collection = collections.get(loader, {})
            collection[var] = obj
            collections[loader] = collection
        stubs = {}
        for loader, collection in collections
            PERSISTENT_LOADERS[loader](loader, self.save_path).save_cache(Cache(collection))
            stubs[loader] = collection.keys()
        self._loader.save_cache(Cache(stubs))
```

Invoking would then be:

```python
with persistent_cache("cache", methods={np.array: "npy", pd.dataframe: "arrow"}):
    ...
```

with method="multi" determined from provided methods kwarg
